File: src/main/java/wraith/fabricaeexnihilo/json/recipe/FluidTransformRecipeJson.java
package wraith.fabricaeexnihilo.json.recipe;

import com.google.gson.*;
import wraith.fabricaeexnihilo.api.crafting.FluidIngredient;
import wraith.fabricaeexnihilo.api.crafting.ItemIngredient;
import wraith.fabricaeexnihilo.api.recipes.barrel.FluidTransformRecipe;
import wraith.fabricaeexnihilo.json.BaseJson;
import wraith.fabricaeexnihilo.modules.barrels.modes.BarrelMode;
import wraith.fabricaeexnihilo.modules.barrels.modes.EmptyMode;

import java.lang.reflect.Type;

public final class FluidTransformRecipeJson extends BaseJson<FluidTransformRecipe> {

    private FluidTransformRecipeJson() {}

    public static final FluidTransformRecipeJson INSTANCE = new FluidTransformRecipeJson();

    @Override
    public FluidTransformRecipe deserialize(JsonElement json, Type typeOfT, JsonDeserializationContext context) throws JsonParseException {
        var obj = json.getAsJsonObject();
        return new FluidTransformRecipe(
                FluidIngredient.fromJson(obj.get("inBarrel"), context),
                ItemIngredient.fromJson(obj.get("catalyst"), context),
                BarrelMode.BARREL_MODE_FACTORY(obj.get("result"), context)
        );
    }

    @Override
    public JsonElement serialize(FluidTransformRecipe src, Type typeOfSrc, JsonSerializationContext context) {
        var obj = new JsonObject();
        obj.add("inBarrel", src.inBarrel().toJson(context));
        obj.add("catalyst", src.catalyst().toJson(context));
        if(!(src.result() instanceof EmptyMode)) {
            var result = new JsonObject();
            result.add(src.result().nbtKey(), context.serialize(src.result()));
            obj.add("result", result);
        }
        return obj;
    }

}

File: src/main/kotlin/exnihilofabrico/json/FluidTransformRecipeJson.kt
package exnihilofabrico.json

import com.google.gson.*
import exnihilofabrico.api.crafting.FluidIngredient
import exnihilofabrico.api.crafting.ItemIngredient
import exnihilofabrico.api.recipes.barrel.FluidTransformRecipe
import exnihilofabrico.modules.barrels.modes.EmptyMode
import exnihilofabrico.modules.barrels.modes.barrelModeFactory
import java.lang.reflect.Type

object FluidTransformRecipeJson: JsonDeserializer<FluidTransformRecipe>, JsonSerializer<FluidTransformRecipe> {
    override fun deserialize(json: JsonElement, typeOfT: Type, context: JsonDeserializationContext): FluidTransformRecipe {
        val obj = json.asJsonObject
        return FluidTransformRecipe(
            FluidIngredient.fromJson(obj["inBarrel"], context),
            ItemIngredient.fromJson(obj["catalyst"], context),
            barrelModeFactory(obj["result"], context)
        )
    }

    override fun serialize(src: FluidTransformRecipe, typeOfSrc: Type, context: JsonSerializationContext): JsonElement {
        val obj = JsonObject()
        obj.add("inBarrel", src.inBarrel.toJson(context))
        obj.add("catalyst", src.catalyst.toJson(context))
        if(src.result !is EmptyMode) {
            val result = JsonObject()
            result.add(src.result.tagKey(), context.serialize(src.result))
            obj.add("result", result)
        }
        return obj
    }

}